rllib/algorithms/impala/torch/vtrace_torch_v2.py

```diff
@@ -0,0 +1,166 @@
+from typing import List, Union
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.torch_utils import convert_to_torch_tensor
+
+torch, nn = try_import_torch()
+
+
+def make_time_major(
+    tensor: Union["torch.Tensor", List["torch.Tensor"]],
+    *,
+    trajectory_len: int = None,
+    recurrent_seq_len: int = None,
+    drop_last: bool = False,
+):
+    """Swaps batch and trajectory axis.
+
+    Args:
+        tensor: A tensor or list of tensors to swap the axis of.
+            NOTE: Each tensor must have the shape [B * T] where B is the batch size and
+            T is the trajectory length.
+        trajectory_len: The length of each trajectory being transformed.
+            If None then `recurrent_seq_len` must be set.
+        recurrent_seq_len: Sequence lengths if recurrent.
+            If None then `trajectory_len` must be set.
+        drop_last: A bool indicating whether to drop the last
+            trajectory item.
+
+    Returns:
+        res: A tensor with swapped axes or a list of tensors with
+        swapped axes.
+    """
+    if isinstance(tensor, (list, tuple)):
+        return [
+            make_time_major(_tensor, trajectory_len, recurrent_seq_len, drop_last)
+            for _tensor in tensor
+        ]
+
+    assert (trajectory_len != recurrent_seq_len) and (
+        trajectory_len is None or recurrent_seq_len is None
+    ), "Either trajectory_len or recurrent_seq_len must be set."
+
+    if recurrent_seq_len:
+        B = recurrent_seq_len.shape[0]
+        T = tensor.shape[0] // B
+    else:
+        # Important: chop the tensor into batches at known episode cut
+        # boundaries.
+        # TODO: (sven) this is kind of a hack and won't work for
+        #  batch_mode=complete_episodes.
+        T = trajectory_len
+        B = tensor.shape[0] // T
+    rs = torch.reshape(tensor, [B, T] + list(tensor.shape[1:]))
+
+    # Swap B and T axes.
+    res = torch.transpose(rs, 1, 0)
+
+    if drop_last:
+        return res[:-1]
+    return res
+
+
+def vtrace_torch(
```

Comment:
Note: This code is a blend of the same counterpart in tf and the old torch vtrace code.